### PR TITLE
Refactor pkg/subctl/cmd/ -> pkg/cloud for RHOS

### PR DESCRIPTION
**Pull request by Jaanki**
Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

### CODE REVIEW

File: pkg/cloud/cleanup/rhos.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+				return err
+			}
+
+			return cloud.CleanupAfterSubmariner(status)
+		})
+	if err != nil {
+		return status.Error(err, "Failed to cleanup RHOS cloud")
+	}
+
+	return nil
```
Comment (tpantelis) [inline pkg/cloud/cleanup/rhos.go]:
```suggestion
	return status.Error(err, "Failed to cleanup RHOS cloud")
```

File: pkg/cloud/prepare/rhos.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/reporter"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/cloud"
+	"github.com/submariner-io/submariner-operator/pkg/cloud/rhos"
+)
+
+func RHOS(restConfigProducer *restconfig.Producer, ports cloud.Ports, config *rhos.Config, status reporter.Interface) error {
```
Comment (tpantelis) [inline pkg/cloud/prepare/rhos.go]: To be consistent with the other params passed by reference.

```suggestion
func RHOS(restConfigProducer *restconfig.Producer, ports *cloud.Ports, config *rhos.Config, status reporter.Interface) error {
```

File: pkg/cloud/prepare/rhos.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+				}
+			}
+
+			return cloud.PrepareForSubmariner(input, status)
+		})
+	if err != nil {
+		return status.Error(err, "Failed to prepare RHOS  cloud")
+	}
+
+	return nil
```
Comment (tpantelis) [inline pkg/cloud/prepare/rhos.go]:
```suggestion
	return status.Error(err, "Failed to prepare RHOS cloud")
```

File: pkg/cloud/rhos/rhos.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	Region           string
+	ProjectID        string
+	OcpMetadataFile  string
+	CloudEntry       string
+	GWInstanceType   string
+}
+
+// RunOn runs the given function on RHOS, supplying it with a cloud instance connected to RHOS and a reporter that writes to CLI.
+// The functions makes sure that infraID and region are specified, and extracts the credentials from a secret in order to connect to RHOS.
+func RunOn(restConfigProducer restconfig.Producer, config *Config, status reporter.Interface,
```
Comment (tpantelis) [inline pkg/cloud/rhos/rhos.go]:
```suggestion
func RunOn(restConfigProducer *restconfig.Producer, config *Config, status reporter.Interface,
```

File: pkg/subctl/cmd/cloud/cleanup/rhos.go [hunk 5]
```diff
@@ -20,39 +20,32 @@ package cleanup
...
+		Run: func(cmd *cobra.Command, args []string) {
+			status := cli.NewReporter()
+
+			err := cleanup.RHOS(&restConfigProducer, status)
+			exit.OnError(err)
+		},
 	}
 
-	rhos.AddRHOSFlags(cmd)
+	rhos.AddRHOSFlags(cmd, &rhosConfig)
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/cleanup/rhos.go:66]: `rhosConfig` is unused so we don't need to add the flags.

File: pkg/subctl/cmd/cloud/prepare/rhos.go [hunk 6]
```diff
@@ -19,67 +19,53 @@ limitations under the License.
...
+	"github.com/submariner-io/submariner-operator/pkg/cloud/prepare"
+	cloudrhos "github.com/submariner-io/submariner-operator/pkg/cloud/rhos"
 	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/cloud/rhos"
 )
 
+var rhosConfig cloudrhos.Config
+
 // newRHOSPrepareCommand returns a new cobra.Command used to prepare a cloud infrastructure.
-func newRHOSPrepareCommand() *cobra.Command {
+func newRHOSPrepareCommand(restConfigProducer *restconfig.Producer, ports cloud.Ports) *cobra.Command {
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/prepare/rhos.go]:
```suggestion
func newRHOSPrepareCommand(restConfigProducer *restconfig.Producer, ports *cloud.Ports) *cobra.Command {
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/cloud/cleanup/rhos.go (hunk 5 above)
Comment (Jaanki) [reply pkg/subctl/cmd/cloud/cleanup/rhos.go:66]: It is needed by `ADDRHOSFlags` because `prepare` needs to pass it. So `cleanup` passes an empty struct.

File: pkg/subctl/cmd/cloud/cleanup/rhos.go [hunk 7]
```diff
@@ -20,39 +20,30 @@ package cleanup
...
+		Run: func(cmd *cobra.Command, args []string) {
+			status := cli.NewReporter()
+
+			err := cleanup.RHOS(&restConfigProducer, status)
+			exit.OnError(err)
+		},
 	}
 
-	rhos.AddRHOSFlags(cmd)
+	rhos.AddRHOSFlags(cmd, &cloudrhos.Config{})
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/cleanup/rhos.go]: This has no affect as the flags are never used.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/cloud/cleanup/rhos.go (hunk 5 above)
Comment (Jaanki) [reply pkg/subctl/cmd/cloud/cleanup/rhos.go:66]: These flags are needed by `rhos.RunOn` and hence `rhosConfig` is populated with appropriate values.

File: pkg/subctl/cmd/cloud/cleanup/rhos.go [hunk 8]
```diff
@@ -20,39 +20,49 @@ package cleanup
...
-		// nolint:wrapcheck // No need to wrap errors here
-		func(cloud api.Cloud, gwDeployer api.GatewayDeployer, status reporter.Interface) error {
-			err := gwDeployer.Cleanup(status)
-			if err != nil {
-				return err
+				exit.OnErrorWithMessage(err, "Failed to read RHOS Cluster information from OCP metadata file")
+			} else {
+				cmdcloud.ExpectFlag(cmdcloud.InfraIDFlag, rhosConfig.InfraID)
+				cmdcloud.ExpectFlag(cmdcloud.RegionFlag, rhosConfig.Region)
+				cmdcloud.ExpectFlag(cmdcloud.ProjectIDFlag, rhosConfig.ProjectID)
```
Comment (Jaanki) [inline pkg/subctl/cmd/cloud/cleanup/rhos.go:54]: I want to reuse this section between both `prepare` and `cleanup`. But given the urgency to merge this, it can be refactored in separate commit later on.

File: pkg/subctl/cmd/cloud/prepare/rhos.go [hunk 9]
```diff
@@ -19,67 +19,54 @@ limitations under the License.
...
-func prepareRHOS(cmd *cobra.Command, args []string) {
-	gwPorts := []api.PortSpec{
-		{Port: nattPort, Protocol: "udp"},
-		{Port: natDiscoveryPort, Protocol: "udp"},
+				exit.OnErrorWithMessage(err, "Failed to read RHOS Cluster information from OCP metadata file")
+			} else {
+				utils.ExpectFlag(utils.InfraIDFlag, rhosConfig.InfraID)
+				utils.ExpectFlag(utils.RegionFlag, rhosConfig.Region)
+				utils.ExpectFlag(utils.ProjectIDFlag, rhosConfig.ProjectID)
+			}
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/prepare/rhos.go]: This code is duplicated with prepare so should be put in a function but we can do that when we put this under _cmd/subctl_.

File: pkg/subctl/cmd/cloud/utils/utils.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	RegionFlag    = "region"
+	ProjectIDFlag = "project-id"
+)
+
+// ExpectFlag exits with an error if the flag value is empty.
+func ExpectFlag(flag, value string) {
+	if value == "" {
+		exit.WithMessage(fmt.Sprintf("You must specify the %v flag", flag))
+	}
+}
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/utils/utils.go]: Why did you add this file? `ExpectFlag` is already defined [here](https://github.com/submariner-io/submariner-operator/blob/0c4ef6c9942718116c8ca46983e03e079c000dbb/pkg/subctl/cmd/utils/utils.go#L63) and the constants are already defined elsewhere.

File: pkg/subctl/cmd/cloud/cloud.go [hunk 11]
```diff
@@ -41,3 +49,10 @@ func init() {
 	restConfigProducer.AddKubeContextFlag(cloudCmd)
 	cmd.AddToRootCommand(cloudCmd)
 }
+
+// ExpectFlag exits with an error if the flag value is empty.
+func ExpectFlag(flag, value string) {
+	if value == "" {
+		exit.WithMessage(fmt.Sprintf("You must specify the %v flag", flag))
+	}
+}
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/cloud.go]: This function is already defined [here](https://github.com/submariner-io/submariner-operator/blob/0c4ef6c9942718116c8ca46983e03e079c000dbb/pkg/subctl/cmd/utils/utils.go#L63).

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/cloud/prepare/rhos.go (hunk 9 above)
Comment (Jaanki) [reply pkg/subctl/cmd/cloud/prepare/rhos.go]: Agree. Right now since bith `prepare` and `cleanup` are in separate packages I duplcated the code. Once moved to `mcd/subctl/` it would be easier to share the code.

File: pkg/subctl/cmd/cloud/utils/utils.go (hunk 10 above)
Comment (Jaanki) [reply pkg/subctl/cmd/cloud/utils/utils.go]: Yupe. Missed that.

File: pkg/subctl/cmd/cloud/cloud.go (hunk 11 above)
Comment (Jaanki) [reply pkg/subctl/cmd/cloud/cloud.go]: Yupe. Removed it.

**Review by tpantelis — APPROVED** [review]

**Review by skitt — APPROVED** [review]

**Comment by skitt** [comment]
Linting is unhappy:

```
pkg/cloud/cleanup/rhos.go:29: File is not `gofumpt`-ed (gofumpt)

pkg/subctl/cmd/cloud/cleanup/rhos.go:21: File is not `gofumpt`-ed (gofumpt)
import (
pkg/subctl/cmd/cloud/cleanup/rhos.go:30: File is not `gofumpt`-ed (gofumpt)
	"os"
pkg/cloud/cleanup/rhos.go:28: unnecessary leading newline (whitespace)
func RHOS(restConfigProducer *restconfig.Producer, config *rhos.Config, status reporter.Interface) error {

pkg/subctl/cmd/cloud/cleanup/rhos.go:22:3: Expected 'o', Found 'g' at pkg/subctl/cmd/cloud/cleanup/rhos.go[line 22,col 3] (gci)
	"github.com/spf13/cobra"
	 ^
```
